### PR TITLE
Update Intersection Widget to Display Slab Thickness

### PR DESCRIPTION
**Pull request by stephencrowell**
Related issue:
* #7083

This pull request updates the Slice Interaction widget for adjusting the thickness of the slab for Thick Slab Reconsturction (TSR).

### CODE REVIEW

**Comment by stephencrowell** [comment]
Currently, the non-interactive version is implemented and updates properly; however, the lines to represent thickness are solid instead of dashed. The image below shows how the lines currently look. This is a known [issue](https://gitlab.kitware.com/vtk/vtk/-/issues/15799) in VTK. There are a few workarounds, but most of them do not work when using 2D actors, properties, and mappers. 

Screenshot from 2023-07-14 12-58-37

**Comment by tao558** [comment]
OK, I've implemented the interactive thick slab reconstruction lines. Attached is a demo of the functionality.

I need to fix some indentation and add a few unit tests. I was planning on using the unit tests in the original [PR](https://github.com/Slicer/Slicer/pull/6124) for the slice intersections as a baseline, but I saw that there weren't any tests added. Testing this kind of thing seems a bit tricky.

@jcfr suggested updating the slice node tests and adding a simple "self" test, but I appreciate any other input you might have @lassoan @pieper 

Thanks

https://github.com/Slicer/Slicer/assets/26352734/a29e3056-0afd-4353-9737-2de06a85058d

File: Libs/MRML/Widgets/qMRMLSliceControllerWidget.cxx [hunk 1]
```diff
@@ -2446,7 +2446,8 @@ void qMRMLSliceControllerWidget::toggleSlabReconstructionInteractive(bool intera
         }
       }
     }
-  // Now mark the nodes as modified
+  // The vtkMRMLSliceIntersectionWidget should observe slice display node modifications
+  // but as a workaround for now, trigger update by modifying all the slice nodes.
```
Comment (jcfr) [inline Libs/MRML/Widgets/qMRMLSliceControllerWidget.cxx:2448]: Should the same workaround be added to `qMRMLSliceControllerWidget::showSlabReconstructionWidget` ?

Comment (tao558) [reply Libs/MRML/Widgets/qMRMLSliceControllerWidget.cxx:2448]: I assumed that the `setSlabReconstructionEnabled` call [here](https://github.com/Slicer/Slicer/blob/ddb55ffda47ef88987073ebe72d70a5353435f8b/Libs/MRML/Widgets/qMRMLSliceControllerWidget.cxx#L2413) would effectively be the same / mark that node as modified, but I may be wrong

**Comment by jcfr** [comment]
To investigate:
* Ensure update of thick slab offset for linked views happen in real time only when the view are hot-linked.
* Confirm assumption that `vtkMRMLSliceIntersectionWidget::ProcessOffsetThickSlabHandle` is slow ...

**Comment by jcfr** [comment]
> re: Ensure update of thick slab offset for linked views happen in real time only when the view are hot-linked.


After further consideration, my previous comment "Ensure update of thick slab offset for linked views happen in real time only when the view are hot-linked." is inaccurate.

To clarify, when we adjust the slice thickness for the red view while interacting with the handle in the green or yellow view, we are modifying properties of the red slice node. Regardless of whether the views are linked or not, the update of the slice viewer is solely dependent on the properties of the individual slice nodes. The "link" feature, as it currently operates, propagates changes to other slice nodes within a given view group, but it does not affect how the slice viewer responds to these changes.

In essence, the linkage status of the views does not impact the real-time update of the slice viewer when adjusting thick slab offset. The viewer's behavior is determined by the properties of the specific slice nodes being modified.

**Comment by jcfr** [comment]
Following today's Slicer weekly meeting:
* The `StartExpensiveUpdate/EndExpensiveUpdate/IsExpensiveUpdate` API is subjective (the meaning of "epensive" is relative) and not the right level of abstraction. Ideally, the widget itself should support the concept of "preview" before effectively requesting the changes to be associated with the viewer.
* Revert the introduction of the "expensive" API and instead keep the behavior and viewer update synchronous and consistent.

**Comment by jcfr** [comment]
The topic has been consolidated and rebased, it is now ready for a final round of review.

**Comment by jcfr** [comment]
Remaining:
* Add `Thick Slab Reconstruction` to "right click" menu associated with slice viewer (see below)
* Rename `Show Slab Reconstruction` to `Show Thick Slab Reconstruction`
* In controller, ensure option are disabled if `Thick Slab Reconstruction` is disabled



image

File: Libs/MRML/Core/vtkMRMLSliceNode.h [hunk 2]
```diff
@@ -468,6 +468,7 @@ class VTK_MRML_EXPORT vtkMRMLSliceNode : public vtkMRMLAbstractViewNode
   ///    LabelOutlineFlag - broadcast outlining the labelmaps
   ///    SliceVisibleFlag = broadcast display of slice in 3D
   ///    ResetOrientationFlag = broadcast a reset to default orientation to all linked viewers
+  ///    UpdateSlabReconstructionThicknessFlag = broadcast updating the slab reconstruction offset
```
Comment (jcfr) [inline Libs/MRML/Core/vtkMRMLSliceNode.h:471]:
```suggestion
  ///    UpdateSlabReconstructionThicknessFlag = broadcast updating the slab reconstruction thickness
```

File: Libs/MRML/DisplayableManager/vtkMRMLSliceIntersectionInteractionRepresentation.h [hunk 3]
```diff
@@ -96,6 +96,8 @@ class VTK_MRML_DISPLAYABLEMANAGER_EXPORT vtkMRMLSliceIntersectionInteractionRepr
     /// Compute slice intersection point between red, green and yellow slice nodes
     void ComputeSliceIntersectionPoint();
 
+    bool DistanceFromSliceLineToPoint(vtkMRMLSliceNode* intersectingSlice, double point_RAS[3], double& distance);
```
Comment (jcfr) [inline Libs/MRML/DisplayableManager/vtkMRMLSliceIntersectionInteractionRepresentation.h:99]: `DistanceFromSliceLineToPoint` -> `DistanceFromSliceIntersectionToPoint`

**Comment by jcfr** [comment]
| Slice Controller (Default) | Slice Controller (Slab Enabled) | Slice Controller (Slab Enabled + Interactive) |
|--|--|--|
| image | image | image |

| Slice Viewer (Handle Interaction )| Slice Viewer (Right-click context menu) |
|--|--|
| image | image |

**Comment by jcfr** [comment]
Windows test build reported:

```
D:\D\P\S-0\Libs\MRML\DisplayableManager\vtkMRMLSliceIntersectionRepresentation2D.cxx(548,34): error C2065: 'M_PI_2': un
declared identifier [D:\D\P\S-0-build\Slicer-build\Libs\MRML\DisplayableManager\MRMLDisplayableManager.vcxproj]
```

**Review by sjh26 — APPROVED** [review]
lgtm

**Pull request merged**